File: hasCycle/fastestSolution.cpp
// link: https://leetcode.com/problems/linked-list-cycle/

class Solution {
public:
    bool hasCycle(ListNode *head) {
        if(head == NULL)
   	    	return false;
        ListNode *slow=head,*fast = head;
    	    while(slow && fast && fast->next)
    	    {
    	    	
			slow = slow ->next;
    	    fast = fast->next->next;
    	    if(slow == fast)
    	    	return true;
    	    else
    	    	continue;
    	    
    	}
    	return false;
        
        
    }
};

File: minimumLengthEncoding/fastestSolution.java
// link: https://leetcode.com/problems/short-encoding-of-words/



class Solution {
    class Trie {
        Trie[] next = null;
    }
    

    private int visit(String s, Trie node) {
        boolean newBranch = false;
        int created = 0, length = s.length();
        for (int i = length - 1; i >= 0; i--) {   
            boolean newLevel = false;
            int idx = s.charAt(i) - 'a';
            if (null == node.next) {
                newLevel = true;
                node.next = new Trie[26];
            }

            if (null == node.next[idx])  {
                if (!newLevel) newBranch = true;
                node.next[idx] = new Trie();
                created++;
            }            
            node = node.next[idx];
        }

        return newBranch ? length + 1 : created;
    }
    
    public int minimumLengthEncoding(String[] words) {
        Trie root = new Trie();
        root.next = new Trie[26];
        int ret = 0;
        for (String s:words) {
            ret += visit(s, root);
        }
        return ret;
    }
}
